File: src/main/java/kr/devport/api/domain/common/webhook/CrawlerWebhookController.java
package kr.devport.api.domain.common.webhook;

import jakarta.validation.Valid;
import kr.devport.api.domain.common.webhook.dto.CrawlerJobCompletedRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

/**
 * Webhook endpoint for crawler-initiated cache invalidation.
 * Receives job-completed events from external crawler system.
 */
@RestController
@RequestMapping("/api/webhooks/crawler")
@RequiredArgsConstructor
@Slf4j
public class CrawlerWebhookController {
    
    private final CrawlerWebhookService crawlerWebhookService;
    
    /**
     * Receives crawler job-completed webhook.
     * 
     * Validates HMAC signature before processing.
     * Returns 202 Accepted for idempotent retry safety.
     * 
     * @param request Job completion payload with signature
     * @param rawPayload Raw JSON for signature verification
     * @return 202 on success, 401 on invalid signature, 400 on validation error
     */
    @PostMapping("/job-completed")
    public ResponseEntity<Map<String, Object>> handleJobCompleted(
            @Valid @RequestBody CrawlerJobCompletedRequest request,
            @RequestBody String rawPayload) {
        
        log.debug("Received crawler webhook: jobId={}", request.getJobId());
        
        // Validate HMAC signature
        if (!crawlerWebhookService.validateSignature(rawPayload, request.getSignature())) {
            log.warn("Rejected webhook with invalid signature: jobId={}", request.getJobId());
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(Map.of(
                    "error", "Invalid signature",
                    "job_id", request.getJobId()
                ));
        }
        
        // Process webhook (idempotent)
        crawlerWebhookService.handleJobCompleted(request);
        
        // Return 202 Accepted for retry-safe semantics
        return ResponseEntity.status(HttpStatus.ACCEPTED)
            .body(Map.of(
                "status", "accepted",
                "job_id", request.getJobId(),
                "scope", request.getEffectiveScope()
            ));
    }
}

File: application-api/src/main/kotlin/kr/devport/api/domain/common/webhook/CrawlerWebhookController.kt
package kr.devport.api.domain.common.webhook

import jakarta.validation.Valid
import kr.devport.api.domain.common.webhook.dto.CrawlerJobCompletedRequest
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

/**
 * Webhook endpoint for crawler-initiated cache invalidation. Receives job-completed events from the
 * external crawler system.
 */
@RestController
@RequestMapping("/api/webhooks/crawler")
class CrawlerWebhookController(
    private val crawlerWebhookService: CrawlerWebhookService,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * Receives a crawler job-completed webhook. Validates the HMAC signature before processing and
     * returns 202 Accepted for idempotent retry safety (401 on invalid signature).
     */
    @PostMapping("/job-completed")
    fun handleJobCompleted(
        @Valid @RequestBody request: CrawlerJobCompletedRequest,
        @RequestBody rawPayload: String,
    ): ResponseEntity<Map<String, Any?>> {
        log.debug("Received crawler webhook: jobId={}", request.jobId)

        if (!crawlerWebhookService.validateSignature(rawPayload, request.signature)) {
            log.warn("Rejected webhook with invalid signature: jobId={}", request.jobId)
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(
                mapOf(
                    "error" to "Invalid signature",
                    "job_id" to request.jobId,
                ),
            )
        }

        crawlerWebhookService.handleJobCompleted(request)

        return ResponseEntity.status(HttpStatus.ACCEPTED).body(
            mapOf(
                "status" to "accepted",
                "job_id" to request.jobId,
                "scope" to request.effectiveScope,
            ),
        )
    }
}
